pkg/sql/catalog/replication/reader_catalog.go

```diff
@@ -0,0 +1,270 @@
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/dbdesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/funcdesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/nstree"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/schemadesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+)
+
+// SetupOrAdvanceStandbyReaderCatalog sets up a reader catalog that reads from
+// that tenant as of the passed timestamp -- potentially replacing the current
+// reader catalog if needed -- or advances the current reader catalog to read as
+// of that timestamp if that is possible to do in place by updating the ts and
+// bumping the descriptor versions. In addition to mirroring all user-created
+// catalog entries, this also updates the system.users, roles and role options,
+// and database role settings tables in this cluster to read from those tables in
+// the other cluster (and fails if they have incompatible indexes). Changes to
+// the catalog are atomic.
+func SetupOrAdvanceStandbyReaderCatalog(
+	ctx context.Context,
+	fromID roachpb.TenantID,
+	asOf hlc.Timestamp,
+	descsCol descs.DB,
+	st *cluster.Settings,
+) error {
+	extracted, err := getCatalogForTenantAsOf(ctx, st, descsCol.KV(), fromID, asOf)
+	if err != nil {
+		return err
+	}
+	return descsCol.DescsTxn(
+		ctx, func(ctx context.Context, txn descs.Txn) error {
+			// Descriptors that no longer exist
+			descriptorsUpdated := catalog.DescriptorIDSet{}
+			namespaceUpdated := catalog.DescriptorIDSet{}
+			allExistingDescs, err := txn.Descriptors().GetAll(ctx, txn.KV())
+			if err != nil {
+				return err
+			}
+			// Resolve any existing descriptors within the tenant, which
+			// will be use to compute old values for writing.
+			b := txn.KV().NewBatch()
+			if err := extracted.ForEachDescriptor(func(desc catalog.Descriptor) error {
+				if !shouldSetupForReader(desc.GetID(), desc.GetParentID()) {
+					return nil
+				}
+				// Track this descriptor was updated.
+				descriptorsUpdated.Add(desc.GetID())
+				// If there is an existing descriptor with the same ID, we should
+				// determine the old bytes in storage for the upsert.
+				var existingRawBytes []byte
+				existingDesc, err := txn.Descriptors().MutableByID(txn.KV()).Desc(ctx, desc.GetID())
+				if err == nil {
+					existingRawBytes = existingDesc.GetRawBytesInStorage()
+				} else if errors.Is(err, catalog.ErrDescriptorNotFound) {
+					err = nil
+				} else {
+					return err
+				}
+				// Existing descriptor should never be a system descriptor.
```

Comment:
nit: this comment is slightly misleading and i think the check could be stronger. Could the check just be:
```
existingDesc.GetParentID()  == desc.GetParentID()
```
i.e. we're asserting that all extracted user tables are existing user tables, and all extracted system table are a existing system tables.

Reply:
we should probably also add an assertion that every extracted system table exists in the reader tenant.

Reply:
^ actually, my check above :
```
existingDesc.GetParentID()  == desc.GetParentID()
```
could lead to false positives if the system database id is not hard coded at start up